<launch>

  <arg name="robot" default="orcvio_robot"/>
  <arg name="fixed_frame_id" default="world"/>

  <!-- bag parameters -->
  <arg name="path_bag" default="/media/erl/disk1/orcvio/mesa_rosbags/indemind_journal_with_keypoints.bag" />
  <arg name="visualize_rviz" default="true" />

  <!-- System Manager Nodelet  -->
  <group ns="$(arg robot)">
    <node pkg="nodelet" type="nodelet" name="system"
      args='standalone orcvio/SystemNodelet'
      output="screen">

      <param name="fixed_frame_id" value="$(arg fixed_frame_id)"/>

      <remap from="~imu" to="/indemind/imu"/>
      <remap from="~cam0_image" to="/indemind/left/image_raw"/>

      <param name="config_file_template" value="$(find orcvio)/../../../config/mesa.yaml"/>
      <param name="config_file" value="$(find orcvio)/../../../cache/mesa_exp.yaml"/>
      <param name="result_file" value="$(find orcvio)/../../../cache/temp_rmse.txt"/>

      <param name="load_groundtruth_flag" type="int" value="0"/>
      <!-- <param name="use_left_perturbation_flag" type="int" value="1" /> -->

      <param name="publish_tf" type="bool" value="true" />

    </node>

    <!-- Object Mapper Nodelet  -->
    <!-- <node pkg="nodelet" type="nodelet" name="ObjectMapper"
      args="standalone orcvio/ObjectMapperNodelet" > -->
    <node pkg="nodelet" type="nodelet" name="ObjectMapper"
      args="standalone orcvio/ObjectMapperNodelet"
      output="screen">

          <param name="fixed_frame_id" value="$(arg fixed_frame_id)"/>

          <!-- camera extrinsics -->
          <!-- this is from imu to camera  -->
          <rosparam param="T_cam_imu">
              [
              -9.99977773e-01,   6.62112204e-03,  -1.17829995e-03, 6.00055992e-02,
              -6.62388202e-03,  -9.99975083e-01,   2.45725146e-03, 4.28596622e-04,
              -1.16169696e-03,   2.46574830e-03,   9.99996572e-01, 2.06964138e-03,
              0.00000000e+00,   0.00000000e+00,   0.00000000e+00, 1.00000000e+00
              ]
          </rosparam>

          <!-- bag topics -->
          <param name="topic_keypoint"  type="string" value="/starmap/keypoints" />
          <!-- TODO  -->
          <param name="topic_caminfo"  type="string" value="/indemind/left/camera_info" />
          <param name="topic_pose"  type="string" value="/orcvio_robot/system/poseout" />
          <param name="topic_image"  type="string" value="/indemind/left/image_raw" />

          <param name="max_object_feature_track"             type="int"    value="50" />
          <param name="min_object_feature_track"               type="int"    value="5" />

          <!-- object feature -->
          <rosparam command="load" file="$(find orcvio)/../../../config/object_feat_mesa.yaml"/>
          <!-- <param name="load_gt_object_info_flag"  type="bool" value="true" /> -->
          <param name="load_gt_object_info_flag"  type="bool" value="false" />

          <!-- <param name="to_color_image_flag"  type="bool" value="true" /> -->
          <param name="to_color_image_flag"  type="bool" value="false" />
          
          <param name="use_unity_dataset_flag"  type="bool" value="false" />
          
          <param name="use_bbox_only_flag"  type="bool" value="false" />
          <!-- <param name="use_bbox_only_flag"  type="bool" value="true" /> -->

          <!-- this has to be consistent with VIO config  -->
          <!-- <param name="use_left_perturbation_flag"  type="bool" value="true" /> -->
          <param name="use_left_perturbation_flag"  type="bool" value="false" />

          <!-- <param name="use_new_bbox_residual_flag"  type="bool" value="false" /> -->
          <param name="use_new_bbox_residual_flag"  type="bool" value="true" />

          <param name="result_dir_path_2d_bbox" value="$(find orcvio)/../../../cache/object_2d_bbox_info.txt"/>

    </node>

  </group>

    <!-- play the dataset -->
    <node pkg="rosbag" type="play" name="rosbag" args="$(arg path_bag)" required="true"/>
    
    <!-- open rviz  -->
    <group if="$(arg visualize_rviz)" >
      <node type="rviz" name="rviz" pkg="rviz" args="-d $(find orcvio)/../../../ros_wrapper/src/orcvio/launch/rviz/orcvio_mesa_object.rviz" />
    </group>

</launch>